Vector Hive sync error during write

### Bug Description

**What happened:**

**What you expected:**

**Steps to reproduce:**
```
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

// E2E test: VECTOR type with Hive sync via SQL CREATE TABLE path.
// NOTE: The VECTOR(n) DDL parser lives on HUDI-18360 and is NOT yet on master.
// Until that lands, this script will fail to parse; the corresponding test is
// @Disabled. Once the parser is merged the test can be re-enabled without
// changes to this file.
spark.sql("""
  CREATE TABLE vector_test (
    id LONG,
    name STRING,
    embedding VECTOR(3),
    dt STRING
  ) USING hudi
  PARTITIONED BY (dt)
  LOCATION '/user/hive/warehouse/vector_test'
  TBLPROPERTIES (
    primaryKey = 'id',
    preCombineField = 'name',
    hoodie.datasource.hive_sync.enable = 'true',
    hoodie.datasource.hive_sync.database = 'default',
    hoodie.datasource.hive_sync.table = 'vector_test',
    hoodie.datasource.hive_sync.jdbcurl = 'jdbc:hive2://hiveserver:10000/',
    hoodie.datasource.hive_sync.mode = 'hms',
    hoodie.datasource.hive_sync.partition_fields = 'dt',
    hoodie.datasource.hive_sync.partition_extractor_class = 'org.apache.hudi.hive.MultiPartKeysValueExtractor',
    hoodie.datasource.hive_sync.username = 'hive',
    hoodie.datasource.hive_sync.password = 'hive'
  )
""")

spark.sql("""
  INSERT INTO vector_test VALUES
    (1, 'doc1', array(cast(0.1 as float), cast(0.2 as float), cast(0.3 as float)), '2024-01-01'),
    (2, 'doc2', array(cast(0.4 as float), cast(0.5 as float), cast(0.6 as float)), '2024-01-01')
""")

// Verify table is readable via Spark
spark.sql("select id, name, embedding, dt from vector_test").show(10, false)
println("VECTOR_SQL_TEST_SUCCESS")
```


### Environment

**Hudi version:**
**Query engine:** (Spark/Flink/Trino etc)
Spark3.5 Scala2.12
**Relevant configs:**


### Logs and Stack Trace

```log
org.opentest4j.AssertionFailedError: Expected to find substring 'VECTOR_SQL_TEST_SUCCESS' 1 times, but found 0. Full stdout: 26/04/21 13:58:24 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
Spark context Web UI available at http://adhoc-1:4040
Spark context available as 'sc' (master = local[2], app id = local-1776779906488).
Spark session available as 'spark'.
26/04/21 13:58:28 WARN HiveConf: HiveConf of name hive.metastore.event.db.notification.api.auth does not exist
26/04/21 13:58:28 WARN HiveConf: HiveConf of name hive.server2.webui.enable does not exist
26/04/21 13:58:29 WARN SessionState: METASTORE_FILTER_HOOK will be ignored, since hive.security.authorization.manager is set to instance of HiveAuthorizerFactory.
26/04/21 13:58:31 WARN HoodieTableConfig: hoodie.properties modified to: {hoodie.table.initial.version=9, hoodie.table.keygenerator.type=SIMPLE, hoodie.table.type=COPY_ON_WRITE, hoodie.table.metadata.partitions=files, hoodie.table.partition.fields=dt, hoodie.archivelog.folder=archived, hoodie.table.create.schema={"type":"record","name":"vector_test_record","namespace":"hoodie.vector_test","fields":[{"name":"id","type":["null","long"],"default":null},{"name":"name","type":["null","string"],"default":null},{"name":"embedding","type":["null",{"type":"fixed","name":"vector_float_3","namespace":"","size":12,"logicalType":"vector","dimension":3,"elementType":"FLOAT","storageBacking":"FIXED_BYTES"}],"default":null},{"name":"dt","type":["null","string"],"default":null}]}, hoodie.timeline.layout.version=2, hoodie.timeline.history.path=history, hoodie.table.checksum=341472130, hoodie.datasource.write.drop.partition.columns=false, hoodie.table.format=native, hoodie.table.name=vector_test, hoodie.table.recordkey.fields=id, hoodie.timeline.path=timeline, hoodie.table.ordering.fields=name, hoodie.datasource.write.hive_style_partitioning=true, hoodie.table.metadata.partitions.inflight=, hoodie.database.name=default, hoodie.datasource.write.partitionpath.urlencode=false, hoodie.record.merge.mode=EVENT_TIME_ORDERING, hoodie.table.version=9} (at hdfs://namenode:8020/user/hive/warehouse/vector_test/.hoodie)
26/04/21 13:58:33 WARN HoodieTableConfig: hoodie.properties modified to: {hoodie.table.initial.version=9, hoodie.table.keygenerator.type=SIMPLE, hoodie.table.type=COPY_ON_WRITE, hoodie.table.metadata.partitions=files, hoodie.table.partition.fields=dt, hoodie.archivelog.folder=archived, hoodie.table.create.schema={"type":"record","name":"vector_test_record","namespace":"hoodie.vector_test","fields":[{"name":"id","type":["null","long"],"default":null},{"name":"name","type":["null","string"],"default":null},{"name":"embedding","type":["null",{"type":"fixed","name":"vector_float_3","namespace":"","size":12,"logicalType":"vector","dimension":3,"elementType":"FLOAT","storageBacking":"FIXED_BYTES"}],"default":null},{"name":"dt","type":["null","string"],"default":null}]}, hoodie.timeline.layout.version=2, hoodie.timeline.history.path=history, hoodie.table.checksum=341472130, hoodie.datasource.write.drop.partition.columns=false, hoodie.table.format=native, hoodie.table.name=vector_test, hoodie.table.recordkey.fields=id, hoodie.timeline.path=timeline, hoodie.table.ordering.fields=name, hoodie.datasource.write.hive_style_partitioning=true, hoodie.table.metadata.partitions.inflight=, hoodie.table.index.defs.path=.hoodie/.index_defs/index.json, hoodie.database.name=default, hoodie.datasource.write.partitionpath.urlencode=false, hoodie.record.merge.mode=EVENT_TIME_ORDERING, hoodie.table.version=9} (at hdfs://namenode:8020/user/hive/warehouse/vector_test/.hoodie)
26/04/21 13:58:34 WARN HoodieTableConfig: hoodie.properties modified to: {hoodie.table.initial.version=9, hoodie.table.keygenerator.type=SIMPLE, hoodie.table.type=COPY_ON_WRITE, hoodie.table.metadata.partitions=column_stats,files, hoodie.table.partition.fields=dt, hoodie.archivelog.folder=archived, hoodie.table.create.schema={"type":"record","name":"vector_test_record","namespace":"hoodie.vector_test","fields":[{"name":"id","type":["null","long"],"default":null},{"name":"name","type":["null","string"],"default":null},{"name":"embedding","type":["null",{"type":"fixed","name":"vector_float_3","namespace":"","size":12,"logicalType":"vector","dimension":3,"elementType":"FLOAT","storageBacking":"FIXED_BYTES"}],"default":null},{"name":"dt","type":["null","string"],"default":null}]}, hoodie.timeline.layout.version=2, hoodie.timeline.history.path=history, hoodie.table.checksum=341472130, hoodie.datasource.write.drop.partition.columns=false, hoodie.table.format=native, hoodie.table.name=vector_test, hoodie.table.recordkey.fields=id, hoodie.timeline.path=timeline, hoodie.table.ordering.fields=name, hoodie.datasource.write.hive_style_partitioning=true, hoodie.table.metadata.partitions.inflight=, hoodie.table.index.defs.path=.hoodie/.index_defs/index.json, hoodie.database.name=default, hoodie.datasource.write.partitionpath.urlencode=false, hoodie.record.merge.mode=EVENT_TIME_ORDERING, hoodie.table.version=9} (at hdfs://namenode:8020/user/hive/warehouse/vector_test/.hoodie)
# WARNING: Unable to get Instrumentation. Dynamic Attach failed. You may add this JAR as -javaagent manually, or supply -Djdk.attach.allowAttachSelf
# WARNING: Unable to attach Serviceability Agent. Unable to attach even with module exceptions: [org.apache.hudi.org.openjdk.jol.vm.sa.SASupportException: Sense failed., org.apache.hudi.org.openjdk.jol.vm.sa.SASupportException: Sense failed., org.apache.hudi.org.openjdk.jol.vm.sa.SASupportException: Sense failed.]
26/04/21 13:58:35 WARN HoodieTableConfig: hoodie.properties modified to: {hoodie.table.initial.version=9, hoodie.table.keygenerator.type=SIMPLE, hoodie.table.type=COPY_ON_WRITE, hoodie.table.metadata.partitions=column_stats,files,partition_stats, hoodie.table.partition.fields=dt, hoodie.archivelog.folder=archived, hoodie.table.create.schema={"type":"record","name":"vector_test_record","namespace":"hoodie.vector_test","fields":[{"name":"id","type":["null","long"],"default":null},{"name":"name","type":["null","string"],"default":null},{"name":"embedding","type":["null",{"type":"fixed","name":"vector_float_3","namespace":"","size":12,"logicalType":"vector","dimension":3,"elementType":"FLOAT","storageBacking":"FIXED_BYTES"}],"default":null},{"name":"dt","type":["null","string"],"default":null}]}, hoodie.timeline.layout.version=2, hoodie.timeline.history.path=history, hoodie.table.checksum=341472130, hoodie.datasource.write.drop.partition.columns=false, hoodie.table.format=native, hoodie.table.name=vector_test, hoodie.table.recordkey.fields=id, hoodie.timeline.path=timeline, hoodie.table.ordering.fields=name, hoodie.datasource.write.hive_style_partitioning=true, hoodie.table.metadata.partitions.inflight=, hoodie.table.index.defs.path=.hoodie/.index_defs/index.json, hoodie.database.name=default, hoodie.datasource.write.partitionpath.urlencode=false, hoodie.record.merge.mode=EVENT_TIME_ORDERING, hoodie.table.version=9} (at hdfs://namenode:8020/user/hive/warehouse/vector_test/.hoodie)
26/04/21 13:58:36 WARN WriteMarkersFactory: Timeline-server-based markers are not supported for HDFS: base path hdfs://namenode:8020/user/hive/warehouse/vector_test.  Falling back to direct markers.
26/04/21 13:58:37 WARN WriteMarkersFactory: Timeline-server-based markers are not supported for HDFS: base path hdfs://namenode:8020/user/hive/warehouse/vector_test.  Falling back to direct markers.
26/04/21 13:58:39 WARN WriteMarkersFactory: Timeline-server-based markers are not supported for HDFS: base path hdfs://namenode:8020/user/hive/warehouse/vector_test.  Falling back to direct markers.
26/04/21 13:58:39 WARN HiveConf: HiveConf of name hive.metastore.event.db.notification.api.auth does not exist
26/04/21 13:58:39 WARN HiveConf: HiveConf of name hive.server2.webui.enable does not exist
26/04/21 13:58:39 ERROR HMSDDLExecutor: Failed to update table for vector_test
org.apache.hadoop.hive.metastore.api.InvalidOperationException: The following columns have types incompatible with the existing columns in their respective positions :
embedding
	at org.apache.hadoop.hive.metastore.api.ThriftHiveMetastore$alter_table_with_environment_context_result$alter_table_with_environment_context_resultStandardScheme.read(ThriftHiveMetastore.java:59744) ~[hive-metastore-2.3.9.jar:2.3.9]
	at org.apache.hadoop.hive.metastore.api.ThriftHiveMetastore$alter_table_with_environment_context_result$alter_table_with_environment_context_resultStandardScheme.read(ThriftHiveMetastore.java:59730) ~[hive-metastore-2.3.9.jar:2.3.9]
	at org.apache.hadoop.hive.metastore.api.ThriftHiveMetastore$alter_table_with_environment_context_result.read(ThriftHiveMetastore.java:59672) ~[hive-metastore-2.3.9.jar:2.3.9]
	at org.apache.thrift.TServiceClient.receiveBase(TServiceClient.java:88) ~[libthrift-0.12.0.jar:0.12.0]
	at org.apache.hadoop.hive.metastore.api.ThriftHiveMetastore$Client.recv_alter_table_with_environment_context(ThriftHiveMetastore.java:1693) ~[hive-metastore-2.3.9.jar:2.3.9]
	at org.apache.hadoop.hive.metastore.api.ThriftHiveMetastore$Client.alter_table_with_environment_context(ThriftHiveMetastore.java:1677) ~[hive-metastore-2.3.9.jar:2.3.9]
	at org.apache.hadoop.hive.metastore.HiveMetaStoreClient.alter_table_with_environmentContext(HiveMetaStoreClient.java:373) ~[hive-metastore-2.3.9.jar:2.3.9]
	at org.apache.hadoop.hive.ql.metadata.SessionHiveMetaStoreClient.alter_table_with_environmentContext(SessionHiveMetaStoreClient.java:322) ~[hive-exec-2.3.9-core.jar:2.3.9]
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:?]
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:?]
	at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:?]
	at java.lang.reflect.Method.invoke(Method.java:566) ~[?:?]
	at org.apache.hadoop.hive.metastore.RetryingMetaStoreClient.invoke(RetryingMetaStoreClient.java:173) ~[hive-metastore-2.3.9.jar:2.3.9]
	at com.sun.proxy.$Proxy40.alter_table_with_environmentContext(Unknown Source) ~[?:?]
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:?]
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:?]
	at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:?]
	at java.lang.reflect.Method.invoke(Method.java:566) ~[?:?]
	at org.apache.hadoop.hive.metastore.HiveMetaStoreClient$SynchronizedHandler.invoke(HiveMetaStoreClient.java:2327) ~[hive-metastore-2.3.9.jar:2.3.9]
	at com.sun.proxy.$Proxy40.alter_table_with_environmentContext(Unknown Source) ~[?:?]
	at org.apache.hudi.hive.ddl.HMSDDLExecutor.updateTableDefinition(HMSDDLExecutor.java:157) ~[hoodie-spark-bundle.jar:1.2.0-SNAPSHOT]
	at org.apache.hudi.hive.HoodieHiveSyncClient.updateTableSchema(HoodieHiveSyncClient.java:338) ~[hoodie-spark-bundle.jar:1.2.0-SNAPSHOT]
	at org.apache.hudi.hive.HiveSyncTool.syncSchema(HiveSyncTool.java:439) ~[hoodie-spark-bundle.jar:1.2.0-SNAPSHOT]
	at org.apache.hudi.hive.HiveSyncTool.syncHoodieTable(HiveSyncTool.java:261) ~[hoodie-spark-bundle.jar:1.2.0-SNAPSHOT]
	at org.apache.hudi.hive.HiveSyncTool.doSync(HiveSyncTool.java:195) ~[hoodie-spark-bundle.jar:1.2.0-SNAPSHOT]
	at org.apache.hudi.hive.HiveSyncTool.syncHoodieTable(HiveSyncTool.java:180) ~[hoodie-spark-bundle.jar:1.2.0-SNAPSHOT]
	at org.apache.hudi.sync.common.util.SyncUtilHelpers.runHoodieMetaSync(SyncUtilHelpers.java:104) ~[hoodie-spark-bundle.jar:1.2.0-SNAPSHOT]
	at org.apache.hudi.sync.common.util.SyncUtilHelpers.runHoodieMetaSync(SyncUtilHelpers.java:72) ~[hoodie-spark-bundle.jar:1.2.0-SNAPSHOT]
	at org.apache.hudi.HoodieSparkSqlWriterInternal.$anonfun$metaSync$2(HoodieSparkSqlWriter.scala:934) ~[hoodie-spark-bundle.jar:1.2.0-SNAPSHOT]
	at scala.collection.mutable.HashSet.foreach(HashSet.scala:79) ~[scala-library-2.12.18.jar:?]
	at org.apache.hudi.HoodieSparkSqlWriterInternal.metaSync(HoodieSparkSqlWriter.scala:932) ~[hoodie-spark-bundle.jar:1.2.0-SNAPSHOT]
	at org.apache.hudi.HoodieSparkSqlWriterInternal.commitAndPerformPostOperations(HoodieSparkSqlWriter.scala:1035) ~[hoodie-spark-bundle.jar:1.2.0-SNAPSHOT]
	at org.apache.hudi.HoodieSparkSqlWriterInternal.writeInternal(HoodieSparkSqlWriter.scala:559) ~[hoodie-spark-bundle.jar:1.2.0-SNAPSHOT]
	at org.apache.hudi.HoodieSparkSqlWriterInternal.$anonfun$write$1(HoodieSparkSqlWriter.scala:187) ~[hoodie-spark-bundle.jar:1.2.0-SNAPSHOT]
	at org.apache.hudi.HoodieSparkSqlWriterInternal.write(HoodieSparkSqlWriter.scala:205) ~[hoodie-spark-bundle.jar:1.2.0-SNAPSHOT]
	at org.apache.hudi.HoodieSparkSqlWriter$.write(HoodieSparkSqlWriter.scala:127) ~[hoodie-spark-bundle.jar:1.2.0-SNAPSHOT]
	at org.apache.spark.sql.hudi.command.InsertIntoHoodieTableCommand$.run(InsertIntoHoodieTableCommand.scala:120) ~[hoodie-spark-bundle.jar:1.2.0-SNAPSHOT]
	at org.apache.spark.sql.hudi.command.InsertIntoHoodieTableCommand.run(InsertIntoHoodieTableCommand.scala:73) ~[hoodie-spark-bundle.jar:1.2.0-SNAPSHOT]
	at org.apache.spark.sql.execution.command.DataWritingCommandExec.sideEffectResult$lzycompute(commands.scala:113) ~[spark-sql_2.12-3.5.3.jar:3.5.3]
	at org.apache.spark.sql.execution.command.DataWritingCommandExec.sideEffectResult(commands.scala:111) ~[spark-sql_2.12-3.5.3.jar:3.5.3]
	at org.apache.spark.sql.execution.command.DataWritingCommandExec.executeCollect(commands.scala:125) ~[spark-sql_2.12-3.5.3.jar:3.5.3]
	at org.apache.spark.sql.execution.QueryExecution$$anonfun$eagerlyExecuteCommands$1.$anonfun$applyOrElse$1(QueryExecution.scala:107) ~[spark-sql_2.12-3.5.3.jar:3.5.3]
	at org.apache.spark.sql.execution.SQLExecution$.$anonfun$withNewExecutionId$6(SQLExecution.scala:125) ~[spark-sql_2.12-3.5.3.jar:3.5.3]
	at org.apache.spark.sql.execution.SQLExecution$.withSQLConfPropagated(SQLExecution.scala:201) ~[spark-sql_2.12-3.5.3.jar:3.5.3]
	at org.apache.spark.sql.execution.SQLExecution$.$anonfun$withNewExecutionId$1(SQLExecution.scala:108) ~[spark-sql_2.12-3.5.3.jar:3.5.3]
	at org.apache.spark.sql.SparkSession.withActive(SparkSession.scala:900) ~[spark-sql_2.12-3.5.3.jar:3.5.3]
	at org.apache.spark.sql.execution.SQLExecution$.withNewExecutionId(SQLExecution.scala:66) ~[spark-sql_2.12-3.5.3.jar:3.5.3]
	at org.apache.spark.sql.execution.QueryExecution$$anonfun$eagerlyExecuteCommands$1.applyOrElse(QueryExecution.scala:107) ~[spark-sql_2.12-3.5.3.jar:3.5.3]
	at org.apache.spark.sql.execution.QueryExecution$$anonfun$eagerlyExecuteCommands$1.applyOrElse(QueryExecution.scala:98) ~[spark-sql_2.12-3.5.3.jar:3.5.3]
	at org.apache.spark.sql.catalyst.trees.TreeNode.$anonfun$transformDownWithPruning$1(TreeNode.scala:461) ~[spark-catalyst_2.12-3.5.3.jar:3.5.3]
	at org.apache.spark.sql.catalyst.trees.CurrentOrigin$.withOrigin(origin.scala:76) ~[spark-sql-api_2.12-3.5.3.jar:3.5.3]
	at org.apache.spark.sql.catalyst.trees.TreeNode.transformDownWithPruning(TreeNode.scala:461) ~[spark-catalyst_2.12-3.5.3.jar:3.5.3]
	at org.apache.spark.sql.catalyst.plans.logical.LogicalPlan.org$apache$spark$sql$catalyst$plans$logical$AnalysisHelper$$super$transformDownWithPruning(LogicalPlan.scala:32) ~[spark-catalyst_2.12-3.5.3.jar:3.5.3]
	at org.apache.spark.sql.catalyst.plans.logical.AnalysisHelper.transformDownWithPruning(AnalysisHelper.scala:267) ~[spark-catalyst_2.12-3.5.3.jar:3.5.3]
	at org.apache.spark.sql.catalyst.plans.logical.AnalysisHelper.transformDownWithPruning$(AnalysisHelper.scala:263) ~[spark-catalyst_2.12-3.5.3.jar:3.5.3]
	at org.apache.spark.sql.catalyst.plans.logical.LogicalPlan.transformDownWithPruning(LogicalPlan.scala:32) ~[spark-catalyst_2.12-3.5.3.jar:3.5.3]
	at org.apache.spark.sql.catalyst.plans.logical.LogicalPlan.transformDownWithPruning(LogicalPlan.scala:32) ~[spark-catalyst_2.12-3.5.3.jar:3.5.3]
	at org.apache.spark.sql.catalyst.trees.TreeNode.transformDown(TreeNode.scala:437) ~[spark-catalyst_2.12-3.5.3.jar:3.5.3]
	at org.apache.spark.sql.execution.QueryExecution.eagerlyExecuteCommands(QueryExecution.scala:98) ~[spark-sql_2.12-3.5.3.jar:3.5.3]
	at org.apache.spark.sql.execution.QueryExecution.commandExecuted$lzycompute(QueryExecution.scala:85) ~[spark-sql_2.12-3.5.3.jar:3.5.3]
	at org.apache.spark.sql.execution.QueryExecution.commandExecuted(QueryExecution.scala:83) ~[spark-sql_2.12-3.5.3.jar:3.5.3]
	at org.apache.spark.sql.Dataset.<init>(Dataset.scala:220) ~[spark-sql_2.12-3.5.3.jar:3.5.3]
	at org.apache.spark.sql.Dataset$.$anonfun$ofRows$2(Dataset.scala:100) ~[spark-sql_2.12-3.5.3.jar:3.5.3]
	at org.apache.spark.sql.SparkSession.withActive(SparkSession.scala:900) ~[spark-sql_2.12-3.5.3.jar:3.5.3]
	at org.apache.spark.sql.Dataset$.ofRows(Dataset.scala:97) ~[spark-sql_2.12-3.5.3.jar:3.5.3]
	at org.apache.spark.sql.SparkSession.$anonfun$sql$4(SparkSession.scala:691) ~[spark-sql_2.12-3.5.3.jar:3.5.3]
	at org.apache.spark.sql.SparkSession.withActive(SparkSession.scala:900) ~[spark-sql_2.12-3.5.3.jar:3.5.3]
	at org.apache.spark.sql.SparkSession.sql(SparkSession.scala:682) ~[spark-sql_2.12-3.5.3.jar:3.5.3]
	at org.apache.spark.sql.SparkSession.sql(SparkSession.scala:713) ~[spark-sql_2.12-3.5.3.jar:3.5.3]
	at org.apache.spark.sql.SparkSession.sql(SparkSession.scala:744) ~[spark-sql_2.12-3.5.3.jar:3.5.3]
	at $line14.$read$$iw$$iw$$iw$$iw$$iw$$iw$$iw$$iw.<init>(/var/hoodie/ws/docker/demo/sparksql-vector-type-sql.commands:69) ~[scala-library-2.12.18.jar:?]
	at $line14.$read$$iw$$iw$$iw$$iw$$iw$$iw$$iw.<init>(/var/hoodie/ws/docker/demo/sparksql-vector-type-sql.commands:86) ~[scala-library-2.12.18.jar:?]
	at $line14.$read$$iw$$iw$$iw$$iw$$iw$$iw.<init>(/var/hoodie/ws/docker/demo/sparksql-vector-type-sql.commands:88) ~[scala-library-2.12.18.jar:?]
	at $line14.$read$$iw$$iw$$iw$$iw$$iw.<init>(/var/hoodie/ws/docker/demo/sparksql-vector-type-sql.commands:90) ~[scala-library-2.12.18.jar:?]
	at $line14.$read$$iw$$iw$$iw$$iw.<init>(/var/hoodie/ws/docker/demo/sparksql-vector-type-sql.commands:92) ~[scala-library-2.12.18.jar:?]
	at $line14.$read$$iw$$iw$$iw.<init>(/var/hoodie/ws/docker/demo/sparksql-vector-type-sql.commands:94) ~[scala-library-2.12.18.jar:?]
	at $line14.$read$$iw$$iw.<init>(/var/hoodie/ws/docker/demo/sparksql-vector-type-sql.commands:96) ~[scala-library-2.12.18.jar:?]
	at $line14.$read$$iw.<init>(/var/hoodie/ws/docker/demo/sparksql-vector-type-sql.commands:98) ~[scala-library-2.12.18.jar:?]
	at $line14.$read.<init>(/var/hoodie/ws/docker/demo/sparksql-vector-type-sql.commands:100) ~[scala-library-2.12.18.jar:?]
	at $line14.$read$.<init>(/var/hoodie/ws/docker/demo/sparksql-vector-type-sql.commands:104) ~[scala-library-2.12.18.jar:?]
	at $line14.$read$.<clinit>(/var/hoodie/ws/docker/demo/sparksql-vector-type-sql.commands) ~[scala-library-2.12.18.jar:?]
	at $line14.$eval$.$print$lzycompute(/var/hoodie/ws/docker/demo/sparksql-vector-type-sql.commands:7) ~[scala-library-2.12.18.jar:?]
	at $line14.$eval$.$print(/var/hoodie/ws/docker/demo/sparksql-vector-type-sql.commands:6) ~[scala-library-2.12.18.jar:?]
	at $line14.$eval.$print(/var/hoodie/ws/docker/demo/sparksql-vector-type-sql.commands) ~[scala-library-2.12.18.jar:?]
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:?]
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:?]
	at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:?]
	at java.lang.reflect.Method.invoke(Method.java:566) ~[?:?]
	at scala.tools.nsc.interpreter.IMain$ReadEvalPrint.call(IMain.scala:747) ~[scala-compiler-2.12.18.jar:?]
	at scala.tools.nsc.interpreter.IMain$Request.loadAndRun(IMain.scala:1020) ~[scala-compiler-2.12.18.jar:?]
	at scala.tools.nsc.interpreter.IMain.$anonfun$interpret$1(IMain.scala:568) ~[scala-compiler-2.12.18.jar:?]
	at scala.reflect.internal.util.ScalaClassLoader.asContext(ScalaClassLoader.scala:36) ~[scala-reflect-2.12.18.jar:?]
	at scala.reflect.internal.util.ScalaClassLoader.asContext$(ScalaClassLoader.scala:116) ~[scala-reflect-2.12.18.jar:?]
	at scala.reflect.internal.util.AbstractFileClassLoader.asContext(AbstractFileClassLoader.scala:41) ~[scala-reflect-2.12.18.jar:?]
	at scala.tools.nsc.interpreter.IMain.loadAndRunReq$1(IMain.scala:567) ~[scala-compiler-2.12.18.jar:?]
	at scala.tools.nsc.interpreter.IMain.interpret(IMain.scala:594) ~[scala-compiler-2.12.18.jar:?]
	at scala.tools.nsc.interpreter.IMain.interpret(IMain.scala:564) ~[scala-compiler-2.12.18.jar:?]
	at scala.tools.nsc.interpreter.ILoop.$anonfun$pasteCommand$11(ILoop.scala:795) ~[scala-compiler-2.12.18.jar:?]
	at scala.tools.nsc.interpreter.IMain.withLabel(IMain.scala:111) ~[scala-compiler-2.12.18.jar:?]
	at scala.tools.nsc.interpreter.ILoop.interpretCode$1(ILoop.scala:795) ~[scala-compiler-2.12.18.jar:?]
	at scala.tools.nsc.interpreter.ILoop.pasteCommand(ILoop.scala:801) ~[scala-compiler-2.12.18.jar:?]
	at org.apache.spark.repl.SparkILoop.$anonfun$process$8(SparkILoop.scala:177) ~[spark-repl_2.12-3.5.3.jar:3.5.3]
	at org.apache.spark.repl.SparkILoop.$anonfun$process$8$adapted(SparkILoop.scala:176) ~[spark-repl_2.12-3.5.3.jar:3.5.3]
	at scala.collection.immutable.List.foreach(List.scala:431) ~[scala-library-2.12.18.jar:?]
	at org.apache.spark.repl.SparkILoop.loadInitFiles$1(SparkILoop.scala:176) ~[spark-repl_2.12-3.5.3.jar:3.5.3]
	at org.apache.spark.repl.SparkILoop.$anonfun$process$4(SparkILoop.scala:166) ~[spark-repl_2.12-3.5.3.jar:3.5.3]
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23) ~[scala-library-2.12.18.jar:?]
	at scala.tools.nsc.interpreter.ILoop.$anonfun$mumly$1(ILoop.scala:166) ~[scala-compiler-2.12.18.jar:?]
	at scala.tools.nsc.interpreter.IMain.beQuietDuring(IMain.scala:206) ~[scala-compiler-2.12.18.jar:?]
	at scala.tools.nsc.interpreter.ILoop.mumly(ILoop.scala:163) ~[scala-compiler-2.12.18.jar:?]
	at org.apache.spark.repl.SparkILoop.loopPostInit$1(SparkILoop.scala:153) ~[spark-repl_2.12-3.5.3.jar:3.5.3]
	at org.apache.spark.repl.SparkILoop.$anonfun$process$10(SparkILoop.scala:221) ~[spark-repl_2.12-3.5.3.jar:3.5.3]
	at org.apache.spark.repl.SparkILoop.withSuppressedSettings$1(SparkILoop.scala:189) [spark-repl_2.12-3.5.3.jar:3.5.3]
	at org.apache.spark.repl.SparkILoop.startup$1(SparkILoop.scala:201) [spark-repl_2.12-3.5.3.jar:3.5.3]
	at org.apache.spark.repl.SparkILoop.process(SparkILoop.scala:236) [spark-repl_2.12-3.5.3.jar:3.5.3]
	at org.apache.spark.repl.Main$.doMain(Main.scala:78) [spark-repl_2.12-3.5.3.jar:3.5.3]
	at org.apache.spark.repl.Main$.main(Main.scala:58) [spark-repl_2.12-3.5.3.jar:3.5.3]
	at org.apache.spark.repl.Main.main(Main.scala) [spark-repl_2.12-3.5.3.jar:3.5.3]
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:?]
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:?]
	at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:?]
	at java.lang.reflect.Method.invoke(Method.java:566) ~[?:?]
	at org.apache.spark.deploy.JavaMainApplication.start(SparkApplication.scala:52) [spark-core_2.12-3.5.3.jar:3.5.3]
	at org.apache.spark.deploy.SparkSubmit.org$apache$spark$deploy$SparkSubmit$$runMain(SparkSubmit.scala:1029) [spark-core_2.12-3.5.3.jar:3.5.3]
	at org.apache.spark.deploy.SparkSubmit.doRunMain$1(SparkSubmit.scala:194) [spark-core_2.12-3.5.3.jar:3.5.3]
	at org.apache.spark.deploy.SparkSubmit.submit(SparkSubmit.scala:217) [spark-core_2.12-3.5.3.jar:3.5.3]
	at org.apache.spark.deploy.SparkSubmit.doSubmit(SparkSubmit.scala:91) [spark-core_2.12-3.5.3.jar:3.5.3]
	at org.apache.spark.deploy.SparkSubmit$$anon$2.doSubmit(SparkSubmit.scala:1120) [spark-core_2.12-3.5.3.jar:3.5.3]
	at org.apache.spark.deploy.SparkSubmit$.main(SparkSubmit.scala:1129) [spark-core_2.12-3.5.3.jar:3.5.3]
	at org.apache.spark.deploy.SparkSubmit.main(SparkSubmit.scala) [spark-core_2.12-3.5.3.jar:3.5.3]
org.apache.hudi.exception.HoodieMetaSyncException: Could not sync using the meta sync class org.apache.hudi.hive.HiveSyncTool
  at org.apache.hudi.sync.common.util.SyncUtilHelpers.runHoodieMetaSync(SyncUtilHelpers.java:106)
  at org.apache.hudi.sync.common.util.SyncUtilHelpers.runHoodieMetaSync(SyncUtilHelpers.java:72)
  at org.apache.hudi.HoodieSparkSqlWriterInternal.$anonfun$metaSync$2(HoodieSparkSqlWriter.scala:934)
  at scala.collection.mutable.HashSet.foreach(HashSet.scala:79)
  at org.apache.hudi.HoodieSparkSqlWriterInternal.metaSync(HoodieSparkSqlWriter.scala:932)
  at org.apache.hudi.HoodieSparkSqlWriterInternal.commitAndPerformPostOperations(HoodieSparkSqlWriter.scala:1035)
  at org.apache.hudi.HoodieSparkSqlWriterInternal.writeInternal(HoodieSparkSqlWriter.scala:559)
  at org.apache.hudi.HoodieSparkSqlWriterInternal.$anonfun$write$1(HoodieSparkSqlWriter.scala:187)
  at org.apache.hudi.HoodieSparkSqlWriterInternal.write(HoodieSparkSqlWriter.scala:205)
  at org.apache.hudi.HoodieSparkSqlWriter$.write(HoodieSparkSqlWriter.scala:127)
  at org.apache.spark.sql.hudi.command.InsertIntoHoodieTableCommand$.run(InsertIntoHoodieTableCommand.scala:120)
  at org.apache.spark.sql.hudi.command.InsertIntoHoodieTableCommand.run(InsertIntoHoodieTableCommand.scala:73)
  at org.apache.spark.sql.execution.command.DataWritingCommandExec.sideEffectResult$lzycompute(commands.scala:113)
  at org.apache.spark.sql.execution.command.DataWritingCommandExec.sideEffectResult(commands.scala:111)
  at org.apache.spark.sql.execution.command.DataWritingCommandExec.executeCollect(commands.scala:125)
  at org.apache.spark.sql.execution.QueryExecution$$anonfun$eagerlyExecuteCommands$1.$anonfun$applyOrElse$1(QueryExecution.scala:107)
  at org.apache.spark.sql.execution.SQLExecution$.$anonfun$withNewExecutionId$6(SQLExecution.scala:125)
  at org.apache.spark.sql.execution.SQLExecution$.withSQLConfPropagated(SQLExecution.scala:201)
  at org.apache.spark.sql.execution.SQLExecution$.$anonfun$withNewExecutionId$1(SQLExecution.scala:108)
  at org.apache.spark.sql.SparkSession.withActive(SparkSession.scala:900)
  at org.apache.spark.sql.execution.SQLExecution$.withNewExecutionId(SQLExecution.scala:66)
  at org.apache.spark.sql.execution.QueryExecution$$anonfun$eagerlyExecuteCommands$1.applyOrElse(QueryExecution.scala:107)
  at org.apache.spark.sql.execution.QueryExecution$$anonfun$eagerlyExecuteCommands$1.applyOrElse(QueryExecution.scala:98)
  at org.apache.spark.sql.catalyst.trees.TreeNode.$anonfun$transformDownWithPruning$1(TreeNode.scala:461)
  at org.apache.spark.sql.catalyst.trees.CurrentOrigin$.withOrigin(origin.scala:76)
  at org.apache.spark.sql.catalyst.trees.TreeNode.transformDownWithPruning(TreeNode.scala:461)
  at org.apache.spark.sql.catalyst.plans.logical.LogicalPlan.org$apache$spark$sql$catalyst$plans$logical$AnalysisHelper$$super$transformDownWithPruning(LogicalPlan.scala:32)
  at org.apache.spark.sql.catalyst.plans.logical.AnalysisHelper.transformDownWithPruning(AnalysisHelper.scala:267)
  at org.apache.spark.sql.catalyst.plans.logical.AnalysisHelper.transformDownWithPruning$(AnalysisHelper.scala:263)
  at org.apache.spark.sql.catalyst.plans.logical.LogicalPlan.transformDownWithPruning(LogicalPlan.scala:32)
  at org.apache.spark.sql.catalyst.plans.logical.LogicalPlan.transformDownWithPruning(LogicalPlan.scala:32)
  at org.apache.spark.sql.catalyst.trees.TreeNode.transformDown(TreeNode.scala:437)
  at org.apache.spark.sql.execution.QueryExecution.eagerlyExecuteCommands(QueryExecution.scala:98)
  at org.apache.spark.sql.execution.QueryExecution.commandExecuted$lzycompute(QueryExecution.scala:85)
  at org.apache.spark.sql.execution.QueryExecution.commandExecuted(QueryExecution.scala:83)
  at org.apache.spark.sql.Dataset.<init>(Dataset.scala:220)
  at org.apache.spark.sql.Dataset$.$anonfun$ofRows$2(Dataset.scala:100)
  at org.apache.spark.sql.SparkSession.withActive(SparkSession.scala:900)
  at org.apache.spark.sql.Dataset$.ofRows(Dataset.scala:97)
  at org.apache.spark.sql.SparkSession.$anonfun$sql$4(SparkSession.scala:691)
  at org.apache.spark.sql.SparkSession.withActive(SparkSession.scala:900)
  at org.apache.spark.sql.SparkSession.sql(SparkSession.scala:682)
  at org.apache.spark.sql.SparkSession.sql(SparkSession.scala:713)
  at org.apache.spark.sql.SparkSession.sql(SparkSession.scala:744)
  ... 60 elided
Caused by: org.apache.hudi.exception.HoodieException: Got runtime exception when hive syncing vector_test
  at org.apache.hudi.hive.HiveSyncTool.syncHoodieTable(HiveSyncTool.java:183)
  at org.apache.hudi.sync.common.util.SyncUtilHelpers.runHoodieMetaSync(SyncUtilHelpers.java:104)
  ... 103 more
Caused by: org.apache.hudi.hive.HoodieHiveSyncException: failed to sync the table vector_test
  at org.apache.hudi.hive.HiveSyncTool.syncHoodieTable(HiveSyncTool.java:281)
  at org.apache.hudi.hive.HiveSyncTool.doSync(HiveSyncTool.java:195)
  at org.apache.hudi.hive.HiveSyncTool.syncHoodieTable(HiveSyncTool.java:180)
  ... 104 more
Caused by: org.apache.hudi.hive.HoodieHiveSyncException: Failed to update table for vector_test
  at org.apache.hudi.hive.ddl.HMSDDLExecutor.updateTableDefinition(HMSDDLExecutor.java:160)
  at org.apache.hudi.hive.HoodieHiveSyncClient.updateTableSchema(HoodieHiveSyncClient.java:338)
  at org.apache.hudi.hive.HiveSyncTool.syncSchema(HiveSyncTool.java:439)
  at org.apache.hudi.hive.HiveSyncTool.syncHoodieTable(HiveSyncTool.java:261)
  ... 106 more
Caused by: org.apache.hadoop.hive.metastore.api.InvalidOperationException: The following columns have types incompatible with the existing columns in their respective positions :
embedding
  at org.apache.hadoop.hive.metastore.api.ThriftHiveMetastore$alter_table_with_environment_context_result$alter_table_with_environment_context_resultStandardScheme.read(ThriftHiveMetastore.java:59744)
  at org.apache.hadoop.hive.metastore.api.ThriftHiveMetastore$alter_table_with_environment_context_result$alter_table_with_environment_context_resultStandardScheme.read(ThriftHiveMetastore.java:59730)
  at org.apache.hadoop.hive.metastore.api.ThriftHiveMetastore$alter_table_with_environment_context_result.read(ThriftHiveMetastore.java:59672)
  at org.apache.thrift.TServiceClient.receiveBase(TServiceClient.java:88)
  at org.apache.hadoop.hive.metastore.api.ThriftHiveMetastore$Client.recv_alter_table_with_environment_context(ThriftHiveMetastore.java:1693)
  at org.apache.hadoop.hive.metastore.api.ThriftHiveMetastore$Client.alter_table_with_environment_context(ThriftHiveMetastore.java:1677)
  at org.apache.hadoop.hive.metastore.HiveMetaStoreClient.alter_table_with_environmentContext(HiveMetaStoreClient.java:373)
  at org.apache.hadoop.hive.ql.metadata.SessionHiveMetaStoreClient.alter_table_with_environmentContext(SessionHiveMetaStoreClient.java:322)
  at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
  at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
  at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
  at java.base/java.lang.reflect.Method.invoke(Method.java:566)
  at org.apache.hadoop.hive.metastore.RetryingMetaStoreClient.invoke(RetryingMetaStoreClient.java:173)
  at com.sun.proxy.$Proxy40.alter_table_with_environmentContext(Unknown Source)
  at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
  at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
  at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
  at java.base/java.lang.reflect.Method.invoke(Method.java:566)
  at org.apache.hadoop.hive.metastore.HiveMetaStoreClient$SynchronizedHandler.invoke(HiveMetaStoreClient.java:2327)
  at com.sun.proxy.$Proxy40.alter_table_with_environmentContext(Unknown Source)
  at org.apache.hudi.hive.ddl.HMSDDLExecutor.updateTableDefinition(HMSDDLExecutor.java:157)
  ... 109 more
```

--- a/hudi-spark-datasource/hudi-spark-common/src/main/scala/org/apache/spark/sql/hudi/command/CreateHoodieTableCommand.scala
+++ b/hudi-spark-datasource/hudi-spark-common/src/main/scala/org/apache/spark/sql/hudi/command/CreateHoodieTableCommand.scala
@@ -41,7 +41,7 @@ import org.apache.spark.sql.hudi.HoodieSqlCommonUtils.{isUsingHiveCatalog, isUsi
 import org.apache.spark.sql.hudi.command.CreateHoodieTableCommand.validateTableSchema
 import org.apache.spark.sql.hudi.command.exception.HoodieAnalysisException
 import org.apache.spark.sql.internal.StaticSQLConf.SCHEMA_STRING_LENGTH_THRESHOLD
-import org.apache.spark.sql.types.{ArrayType, BinaryType, DataType, MapType, StructField, StructType}
+import org.apache.spark.sql.types.{ArrayType, BinaryType, DataType, MapType, Metadata, StructField, StructType}
 
 import scala.collection.JavaConverters._
 import scala.collection.mutable
@@ -254,7 +254,7 @@ object CreateHoodieTableCommand extends SparkAdapterSupport {
    *   - VECTOR (Hudi custom logical type, exposed in Spark as ArrayType with `hudi_type`
    *     metadata) -> `BinaryType`, matching its on-disk fixed_len_byte_array layout
    *     (RFC-99). Field metadata is preserved so the full logical schema is still
-   *     serialised into `spark.sql.sources.schema.*` TBLPROPERTIES by the caller.
+   *     serialized into `spark.sql.sources.schema.*` TBLPROPERTIES by the caller.
    *
    * Recurses into nested StructType, ArrayType, and MapType so variants embedded in
    * complex types (e.g. `STRUCT<a: VARIANT>`, `ARRAY<VARIANT>`, `MAP<STRING, VARIANT>`)
@@ -264,7 +264,8 @@ object CreateHoodieTableCommand extends SparkAdapterSupport {
     toHiveCompatibleType(schema).asInstanceOf[StructType]
   }
 
-  private def toHiveCompatibleType(dataType: DataType): DataType = dataType match {
+  private def toHiveCompatibleType(dataType: DataType,
+                                   metadata: Metadata = Metadata.empty): DataType = dataType match {
     case dt if sparkAdapter.isVariantType(dt) =>
       // Canonical field order (metadata, value) matches the Parquet spec and Iceberg convention,
       // mirroring HoodieSchema.createVariant().
@@ -272,14 +273,12 @@ object CreateHoodieTableCommand extends SparkAdapterSupport {
         StructField(HoodieSchema.Variant.VARIANT_METADATA_FIELD, BinaryType, nullable = false),
         StructField(HoodieSchema.Variant.VARIANT_VALUE_FIELD, BinaryType, nullable = false)
       ))
+    case _: ArrayType if isVectorType(metadata) =>
+      // VECTOR is exposed in Spark as ArrayType(FloatType) with hudi_type metadata.
+      // HMS must store the column as BINARY matching the on-disk fixed_len_byte_array (RFC-99).
+      BinaryType
     case st: StructType =>
-      StructType(st.fields.map { f =>
-        if (isVectorField(f)) {
-          f.copy(dataType = BinaryType)
-        } else {
-          f.copy(dataType = toHiveCompatibleType(f.dataType))
-        }
-      })
+      StructType(st.fields.map(f => f.copy(dataType = toHiveCompatibleType(f.dataType, f.metadata))))
     case at: ArrayType =>
       at.copy(elementType = toHiveCompatibleType(at.elementType))
     case mt: MapType =>
@@ -289,13 +288,11 @@ object CreateHoodieTableCommand extends SparkAdapterSupport {
     case other => other
   }
 
-  private def isVectorField(field: StructField): Boolean = {
-    // Mirrors the detection pattern in HoodieSparkSchemaConverters.toHoodieTypeNested. The
-    // TYPE_METADATA_FIELD is documented to only hold custom-logical-type descriptors
-    // (VECTOR, BLOB, VARIANT), so parseTypeDescriptor is safe once the key is present.
-    field.metadata.contains(HoodieSchema.TYPE_METADATA_FIELD) &&
+  /** Mirrors the detection in HoodieSparkSchemaConverters.toHoodieTypeNested. */
+  private def isVectorType(metadata: Metadata): Boolean = {
+    metadata.contains(HoodieSchema.TYPE_METADATA_FIELD) &&
       HoodieSchema.parseTypeDescriptor(
-        field.metadata.getString(HoodieSchema.TYPE_METADATA_FIELD)).getType == HoodieSchemaType.VECTOR
+        metadata.getString(HoodieSchema.TYPE_METADATA_FIELD)).getType == HoodieSchemaType.VECTOR
   }
 
   // This code is forked from org.apache.spark.sql.hive.HiveExternalCatalog#tableMetaToTableProps

--- a/hudi-spark-datasource/hudi-spark-common/src/main/scala/org/apache/spark/sql/hudi/command/CreateHoodieTableCommand.scala
+++ b/hudi-spark-datasource/hudi-spark-common/src/main/scala/org/apache/spark/sql/hudi/command/CreateHoodieTableCommand.scala
@@ -19,7 +19,7 @@ package org.apache.spark.sql.hudi.command
 
 import org.apache.hudi.{DataSourceWriteOptions, SparkAdapterSupport}
 import org.apache.hudi.common.model.HoodieTableType
-import org.apache.hudi.common.schema.HoodieSchema
+import org.apache.hudi.common.schema.{HoodieSchema, HoodieSchemaType}
 import org.apache.hudi.common.table.HoodieTableConfig
 import org.apache.hudi.common.util.ConfigUtils
 import org.apache.hudi.exception.{HoodieException, HoodieValidationException}
@@ -249,7 +249,13 @@ object CreateHoodieTableCommand extends SparkAdapterSupport {
 
   /**
    * Converts Spark DataTypes that Hive doesn't support to their physical representations.
-   * Currently handles VariantType (Spark 4.0+) -> `struct<metadata:binary, value:binary>`.
+   * Currently handles:
+   *   - VariantType (Spark 4.0+) -> `struct<metadata:binary, value:binary>`
+   *   - VECTOR (Hudi custom logical type, exposed in Spark as ArrayType with `hudi_type`
+   *     metadata) -> `BinaryType`, matching its on-disk fixed_len_byte_array layout
+   *     (RFC-99). Field metadata is preserved so the full logical schema is still
+   *     serialised into `spark.sql.sources.schema.*` TBLPROPERTIES by the caller.
+   *
    * Recurses into nested StructType, ArrayType, and MapType so variants embedded in
    * complex types (e.g. `STRUCT<a: VARIANT>`, `ARRAY<VARIANT>`, `MAP<STRING, VARIANT>`)
    * are also converted.
@@ -267,7 +273,13 @@ object CreateHoodieTableCommand extends SparkAdapterSupport {
         StructField(HoodieSchema.Variant.VARIANT_VALUE_FIELD, BinaryType, nullable = false)
       ))
     case st: StructType =>
-      StructType(st.fields.map(f => f.copy(dataType = toHiveCompatibleType(f.dataType))))
+      StructType(st.fields.map { f =>
+        if (isVectorField(f)) {
+          f.copy(dataType = BinaryType)
+        } else {
+          f.copy(dataType = toHiveCompatibleType(f.dataType))
+        }
+      })
     case at: ArrayType =>
       at.copy(elementType = toHiveCompatibleType(at.elementType))
     case mt: MapType =>
@@ -275,6 +287,15 @@ object CreateHoodieTableCommand extends SparkAdapterSupport {
         keyType = toHiveCompatibleType(mt.keyType),
         valueType = toHiveCompatibleType(mt.valueType))
     case other => other
+  }
+
+  private def isVectorField(field: StructField): Boolean = {
+    // Mirrors the detection pattern in HoodieSparkSchemaConverters.toHoodieTypeNested. The
+    // TYPE_METADATA_FIELD is documented to only hold custom-logical-type descriptors
+    // (VECTOR, BLOB, VARIANT), so parseTypeDescriptor is safe once the key is present.
+    field.metadata.contains(HoodieSchema.TYPE_METADATA_FIELD) &&
+      HoodieSchema.parseTypeDescriptor(
+        field.metadata.getString(HoodieSchema.TYPE_METADATA_FIELD)).getType == HoodieSchemaType.VECTOR
   }
 
   // This code is forked from org.apache.spark.sql.hive.HiveExternalCatalog#tableMetaToTableProps

--- a/hudi-spark-datasource/hudi-spark/src/test/scala/org/apache/spark/sql/hudi/ddl/TestCreateTable.scala
+++ b/hudi-spark-datasource/hudi-spark/src/test/scala/org/apache/spark/sql/hudi/ddl/TestCreateTable.scala
@@ -37,6 +37,7 @@ import org.apache.spark.sql.catalyst.TableIdentifier
 import org.apache.spark.sql.catalyst.catalog.{CatalogTableType, HoodieCatalogTable}
 import org.apache.spark.sql.functions.{col, concat, expr, lit}
 import org.apache.spark.sql.hudi.HoodieSqlCommonUtils
+import org.apache.spark.sql.hudi.command.CreateHoodieTableCommand
 import org.apache.spark.sql.hudi.common.HoodieSparkSqlTestBase
 import org.apache.spark.sql.hudi.common.HoodieSparkSqlTestBase.{disableComplexKeygenValidation, getLastCommitMetadata}
 import org.apache.spark.sql.types._
@@ -2286,5 +2287,50 @@ class TestCreateTable extends HoodieSparkSqlTestBase {
       assertTrue(embeddingField.metadata.contains(HoodieSchema.TYPE_METADATA_FIELD))
       assertEquals(ArrayType(FloatType, containsNull = false), embeddingField.dataType)
     }
+  }
+
+  test("toHiveCompatibleSchema rewrites VECTOR field to BinaryType and preserves metadata") {
+    // VECTOR's on-disk layout is Parquet fixed_len_byte_array (RFC-99); the HMS column must be
+    // BINARY so Hive-side Parquet reads match the physical type. The logical ArrayType(FloatType)
+    // view is preserved in spark.sql.sources.schema.* TBLPROPERTIES via the retained metadata.
+    val vectorMetadata = new MetadataBuilder()
+      .putString(HoodieSchema.TYPE_METADATA_FIELD, "VECTOR(3)")
+      .putString("comment", "embedding")
+      .build()
+
+    val schema = StructType(Seq(
+      StructField("id", LongType, nullable = false),
+      StructField("name", StringType),
+      StructField(
+        "embedding",
+        ArrayType(FloatType, containsNull = false),
+        nullable = true,
+        metadata = vectorMetadata),
+      StructField("dt", StringType)))
+
+    val hiveSchema = CreateHoodieTableCommand.toHiveCompatibleSchema(schema)
+
+    val embedding = hiveSchema("embedding")
+    assertEquals(BinaryType, embedding.dataType)
+    assertEquals("binary", embedding.dataType.catalogString)
+    assertTrue(embedding.metadata.contains(HoodieSchema.TYPE_METADATA_FIELD))
+    assertEquals("VECTOR(3)", embedding.metadata.getString(HoodieSchema.TYPE_METADATA_FIELD))
+    assertEquals("embedding", embedding.metadata.getString("comment"))
+
+    // Non-VECTOR fields must pass through unchanged.
+    assertEquals(LongType, hiveSchema("id").dataType)
+    assertEquals(StringType, hiveSchema("name").dataType)
+    assertEquals(StringType, hiveSchema("dt").dataType)
+  }
+
+  test("toHiveCompatibleSchema leaves plain ArrayType(FloatType) without VECTOR marker alone") {
+    // Guard against accidentally rewriting any ArrayType(FloatType). Only fields carrying the
+    // VECTOR descriptor in hudi_type metadata should be converted to BinaryType.
+    val schema = StructType(Seq(
+      StructField("id", LongType, nullable = false),
+      StructField("floats", ArrayType(FloatType, containsNull = false))))
+
+    val hiveSchema = CreateHoodieTableCommand.toHiveCompatibleSchema(schema)
+    assertEquals(ArrayType(FloatType, containsNull = false), hiveSchema("floats").dataType)
   }
 }

